### PR TITLE
Improved nullable support

### PR DESCRIPTION
**Pull request by jkurdek**

### CODE REVIEW

File: src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs [hunk 1]
```diff
@@ -111,6 +111,338 @@ private static bool ShouldUseSetter(BindingMode mode, BindableProperty bindableP
...
+                                && (bindableProperty.DefaultBindingMode == BindingMode.OneWayToSource
+                                    || bindableProperty.DefaultBindingMode == BindingMode.TwoWay));
+                }
+            }
+            """,
+            result.GeneratedCode);
+    }
+
+    [Theory]
+    [InlineData(
```
Comment (simonrozsival) [inline src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs:246]: I don't think this is a good use-case for `InlineData`, I suggest using `MemberData` instead.

File: src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs [hunk 2]
```diff
@@ -111,6 +111,338 @@ private static bool ShouldUseSetter(BindingMode mode, BindableProperty bindableP
...
+        label.SetBinding(Label.RotationProperty, static (A a) => a?.B.C);
+
+        namespace MyNamespace
+            {
+                public class A
+                {
+                    public B B { get; set; } = null!;
+                }
+
+                public class B
```
Comment (simonrozsival) [inline src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs:321]: Can you please add one more test case where we have `struct B`? We must avoid checking for nullability (in the case of `public B B { get; set; }`).

**Review by simonrozsival — COMMENTED** [review]
I like the end-to-end tests that you added. I'm not sure if we should transfor the path as it's parsed or if we should just modify the code that produces the C# code. I'm leaning towards the latter, it seems less magic. The code writer should just have some a parameter like `bool considerAllReferenceTypesPotentiallyNullable` and generate a nullability check in the setter for each reference type and always add `?`s after accessing a reference type when constructing the partial getters for the handlers array.

File: src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs (hunk 2 above)
Comment (jkurdek) [reply src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs:321]: Im not sure I got it. You refer to a scenario when we have non-nullable value type on the path, and we want to access one of its properties?

Comment (simonrozsival) [reply src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs:321]: I had something like this in mind:

```c#
// Condtional access operator
#nullable disable
using Microsoft.Maui.Controls;
using MyNamespace;
var label = new Label();
label.SetBinding(Label.RotationProperty, static (A a) => a?.B.C);
namespace MyNamespace
{
    public class A
    {
        public B B { get; set; }
    }
    public struct B
    {
        public C C { get; set; } = null!;
    }
    public class C
    {
    }
}
```
the output would differ in this:
```c#
setter = static (source, value) =>
{
    if (!value.HasValue)
    {
        return;
    }
    if (source is {} p0)
    {
        p0.B.C = value.Value;
    }
};
```

File: src/Controls/src/BindingSourceGen/GeneratorDataModels.cs [hunk 3]
```diff
@@ -54,13 +55,18 @@ public override string ToString()
 
 public sealed record SetterOptions(bool IsWritable, bool AcceptsNullValue = false);
 
-public sealed record MemberAccess(string MemberName) : IPathPart
+public sealed record MemberAccess(string MemberName, bool IsValueType = false) : IPathPart
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/GeneratorDataModels.cs:58]: We need the same for `IndexAccess`, don't we?

File: src/Controls/src/BindingSourceGen/PathParser.cs [hunk 4]
```diff
@@ -41,7 +41,8 @@ internal PathParser(GeneratorSyntaxContext context, bool enabledNullable)
         }
 
         var member = memberAccess.Name.Identifier.Text;
-        IPathPart part = new MemberAccess(member);
+        var IsReferenceType = Context.SemanticModel.GetTypeInfo(memberAccess).Type?.IsReferenceType ?? false;
+        IPathPart part = new MemberAccess(member, !IsReferenceType);
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/PathParser.cs:45]:
```suggestion
        var isReferenceType = Context.SemanticModel.GetTypeInfo(memberAccess).Type?.IsReferenceType ?? false;
        IPathPart part = new MemberAccess(member, !isReferenceType);
```

File: src/Controls/src/BindingSourceGen/PathParser.cs [hunk 5]
```diff
@@ -86,7 +87,8 @@ internal PathParser(GeneratorSyntaxContext context, bool enabledNullable)
     private (EquatableArray<DiagnosticInfo> diagnostics, List<IPathPart> parts) HandleMemberBindingExpression(MemberBindingExpressionSyntax memberBinding)
     {
         var member = memberBinding.Name.Identifier.Text;
-        IPathPart part = new MemberAccess(member);
+        var IsReferenceType = Context.SemanticModel.GetTypeInfo(memberBinding).Type?.IsReferenceType ?? false;
+        IPathPart part = new MemberAccess(member, !IsReferenceType);
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/PathParser.cs:91]:
```suggestion
        var isReferenceType = Context.SemanticModel.GetTypeInfo(memberBinding).Type?.IsReferenceType ?? false;
        IPathPart part = new MemberAccess(member, !isReferenceType);
```

File: src/Controls/src/BindingSourceGen/SetterBuilder.cs [hunk 6]
```diff
@@ -68,6 +67,11 @@ public void AddPart(IPathPart nextPart)
                     AddIsExpression("{}");
                     _expression = AccessExpressionBuilder.Build(_expression, innerPart);
                 }
+                else if (previousPart is MemberAccess memberAccess && !memberAccess.IsMemberValueType && _considerAllReferenceTypesPotentiallyNullable) 
+                {
+                    AddIsExpression("{}");
+                    _expression = AccessExpressionBuilder.Build(_expression, memberAccess);
+                }
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/SetterBuilder.cs:74]: I don't think this is equivalent to the removed condition in `Setter.From`. If this didn't break any existing test though, I need to think a bit if we are missing a test scenario or not.

Comment (jkurdek) [reply src/Controls/src/BindingSourceGen/SetterBuilder.cs:74]: This is still WIP. The previous version contained a bug. When we had nullable source and first access in the path was conditionalAccess it generated `if p0 is {} p1`.

Comment (simonrozsival) [reply src/Controls/src/BindingSourceGen/SetterBuilder.cs:74]: oh, good catch!

**Comment by jkurdek** [comment]
TODO:
- Indexer support
- Review tests
- Add additional tests to SetterBuilderTests
- Refactor

File: src/Controls/src/BindingSourceGen/SetterBuilder.cs [hunk 7]
```diff
@@ -28,32 +25,47 @@ public sealed record Setter(string[] PatternMatchingExpressions, string Assignme
...
         }
 
         public void AddPart(IPathPart nextPart)
         {
-            _previousPart = HandlePreviousPart(nextPart);
+            var newPart = HandleCurrentPart(nextPart);
+            _previousPart = _currentPart;
+            _currentPart = newPart;
+
         }
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/SetterBuilder.cs:56]: Thinking out loud: this needs refactoring, the `nextPart`, `newPart`, `_currentPart`, `_previousPart` naming is confusing.

File: src/Controls/src/BindingSourceGen/PathParser.cs [hunk 8]
```diff
@@ -162,7 +171,9 @@ internal PathParser(GeneratorSyntaxContext context, bool enabledNullable)
         }
 
         var name = GetIndexerName(elementAccessSymbol);
-        IPathPart part = new IndexAccess(name, indexValue);
+        var isReferenceType = typeSymbol?.IsReferenceType ?? false;
+        var isNullableValueType = typeSymbol != null ? BindingGenerationUtilities.IsNullableValueType(typeSymbol) : false;
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/PathParser.cs:175]: There's quite a lot of code duplication going on here and it's getting hard to read. We should consider moving this to a separate method at least.

File: src/Controls/src/BindingSourceGen/BindingCodeWriter.cs [hunk 9]
```diff
@@ -274,20 +274,21 @@ private void AppendSetterAction(SetBindingInvocationDescription binding, string
...
 
 			string nextExpression = "source";
-			bool forceConditonalAccessToNextPart = false;
+			bool forceConditonalAccessToNextPart = !sourceType.IsValueType && considerAllReferenceTypesPotentiallyNullable;
 			foreach (var part in path)
 			{
 				var previousExpression = nextExpression;
 				nextExpression = AccessExpressionBuilder.Build(previousExpression, MaybeWrapInConditionalAccess(part, forceConditonalAccessToNextPart));
-				forceConditonalAccessToNextPart = part is Cast;
+				var isNullableReferenceType = part is MemberAccess memberAccess && !memberAccess.IsValueType;
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/BindingCodeWriter.cs:290]: We need a check for `IndexAccess` here as well + a unit test that covers that case.

**Review by simonrozsival — APPROVED** [review]

**Pull request merged**